### PR TITLE
Singlepass terrain render

### PR DESCRIPTION
**Pull request by lukaspj**
Since we have moved to DX11, we can use texture arrays to send a bunch of textures to the terrain render instead of render it in passes of 3 TerrainMaterials at a time.

Benefits:

 * Performance should improve, as we can do it in 1 drawcall
 * Code is simplified as we don't need the pass-logic
 * Improvements to terrain rendering are significantly easier (such as [changing the blend algorithm](https://github.com/GarageGames/Torque3D/pull/714))
 * We are less limited in number of samplers.

Downsides:

 * All Terrain Textures need to be the exact same resolution. You can't mix a 256x256 texture with a 512x512 texture.
 * All Terrain Textures need to be the same format, PNG vs JPG should be alright but you can't mix e.g. DDS and PNG

The downsides are currently mitigated by the engine attempting to convert and resize the textures on the spot (thanks for the help @Azaezel), which should come as a negligible performance cost, as the TextureArray are constructed very rarely.

Includes #423

### CODE REVIEW

**Comment by lukaspj** [comment]
Draft status:

- [ ]  OpenGL
- [ ] D3D11
- [ ] Polish

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+   if (!found) return;
+   for (U32 idx = 0; idx < mArraySize; ++idx)
+   {
+      if(textureArray[idx].isValid())
+      {
+         D3D11_TEXTURE2D_DESC desc;
+         static_cast<GFXD3D11TextureObject*>(textureArray[idx].getPointer())->get2DTex()->GetDesc(&desc);
+         if (desc.Height != texElementDesc.Height && desc.Width != texElementDesc.Width && desc.Format != texElementDesc.Format)
+         {
+            AssertFatal(true, "GFXD3D11TextureArray::fromTextureArray all textures must have the same size and format");
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: This should be a gentler error

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+         found = true;
+      }
+   }
+   if (!found) return;
+   for (U32 idx = 0; idx < mArraySize; ++idx)
+   {
+      if(textureArray[idx].isValid())
+      {
+         D3D11_TEXTURE2D_DESC desc;
+         static_cast<GFXD3D11TextureObject*>(textureArray[idx].getPointer())->get2DTex()->GetDesc(&desc);
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Dynamic cast instead since we are casting down the inheritance tree

Comment (lukaspj) [reply Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Also should store in variable

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+   createResourceView(texArrayDesc.Format, texArrayDesc.MipLevels, texArrayDesc.BindFlags);
+   //---------------------------------------------------------------------------------------
+}
+
+void GFXD3D11TextureArray::setToTexUnit(U32 tuNum)
+{
+   D3D11DEVICECONTEXT->PSSetShaderResources(tuNum, 1, &mSRView);
+}
+
+
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Unnecessary white space

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+void GFXD3D11TextureArray::createResourceView(DXGI_FORMAT format, U32 numMipLevels, U32 usageFlags)
+{
+   HRESULT hr;
+   //TODO: add MSAA support later.
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Copy paste todo

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+         desc.Format = format;
+
+      desc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURE2DARRAY;
+      desc.Texture2DArray.MostDetailedMip = 0;
+      desc.Texture2DArray.MipLevels = numMipLevels;
+      desc.Texture2DArray.FirstArraySlice = 0;
+      desc.Texture2DArray.ArraySize = mArraySize;
+
+      hr = D3D11DEVICE->CreateShaderResourceView(mTextureArray, &desc, &mSRView);
+      AssertFatal(SUCCEEDED(hr), "CreateShaderResourceView:: failed to create view!");
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Should include class name

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+ID3D11DepthStencilView** GFXD3D11TextureArray::getDSViewPtr()
+{
+   return &mDSView;
+}
+
+void GFXD3D11TextureArray::zombify()
+{
+   // Static cubemaps are handled by D3D
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Copy paste comment

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#ifndef _GFXD3D11TEXTUREARRAY_H_
+#define _GFXD3D11TEXTUREARRAY_H_
+
+#include <dxgiformat.h>
+
+
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.h:6]: Unnecessary white space

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+   ID3D11DepthStencilView** getDSViewPtr();
+
+private:
+   ID3D11ShaderResourceView* mSRView; // for shader resource input
+   ID3D11RenderTargetView* mRTView; // for render targets
+   ID3D11DepthStencilView* mDSView; //render target view for depth stencil
+
+   ID3D11Texture2D* mTextureArray;
+};
+
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.h:54]: Unnecessary white space

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+   void setToTexUnit(U32 tuNum) override;
+
+   void createResourceView(DXGI_FORMAT format, U32 numMipLevels, U32 usageFlags);
+
+   // GFXResource interface
+   void zombify() override;
+   void resurrect() override;
+   void Release() override;
+
+
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.h:35]: Unnecessary white space

File: Engine/source/gfx/D3D11/gfxD3D11TextureObject.cpp [hunk 10]
```diff
@@ -210,8 +210,8 @@ bool GFXD3D11TextureObject::copyToBmp(GBitmap* bmp)
    // check format limitations
    // at the moment we only support RGBA for the source (other 4 byte formats should
    // be easy to add though)
-   AssertFatal(mFormat == GFXFormatR16G16B16A16F || mFormat == GFXFormatR8G8B8A8 || mFormat == GFXFormatR8G8B8A8_LINEAR_FORCE || mFormat == GFXFormatR8G8B8A8_SRGB, "copyToBmp: invalid format");
-   if (mFormat != GFXFormatR16G16B16A16F && mFormat != GFXFormatR8G8B8A8 && mFormat != GFXFormatR8G8B8A8_LINEAR_FORCE && mFormat != GFXFormatR8G8B8A8_SRGB)
+   AssertFatal(mFormat == GFXFormatR16G16B16A16F || mFormat == GFXFormatR8G8B8A8 || mFormat == GFXFormatR8G8B8A8_LINEAR_FORCE || mFormat == GFXFormatR8G8B8A8_SRGB || mFormat == GFXFormatR8G8B8, "copyToBmp: invalid format");
```
Comment (lukaspj) [inline Engine/source/gfx/D3D11/gfxD3D11TextureObject.cpp:213]: These changes are irrelevant

File: Engine/source/gfx/gfxTextureArray.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
+//-----------------------------------------------------------------------------
+// Copyright (c) 2012 GarageGames, LLC
```
Comment (lukaspj) [inline Engine/source/gfx/gfxTextureArray.h:2]: Should the header be changed?

File: Engine/source/gfx/gfxTextureArray.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+class GFXTextureProfile;
+class GFXTextureObject;
+
+class GFXTextureArray : public StrongRefBase, public GFXResource
+{
+public:
+   virtual void fromTextureArray(Vector<GFXTexHandle> textureArray) = 0;
+   virtual void setToTexUnit(U32 tuNum) = 0;
+
```
Comment (lukaspj) [inline Engine/source/gfx/gfxTextureArray.h:52]: Unnecessary white space

File: Engine/source/gfx/gfxTextureArray.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+   virtual void resurrect() = 0;
+   virtual void Release() = 0;
+
+   virtual const String describeSelf() const;
+
+   U32 mArraySize;
+};
+
+
+/// A reference counted handle to a cubemap array resource.
```
Comment (lukaspj) [inline Engine/source/gfx/gfxTextureArray.h]: Copy paste comment

File: Engine/source/gfx/gl/gfxGLTextureArray.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+   }
+
+   bool found = false;
+   U32 baseWidth, baseHeight;
+   bool isCompressed;
+   for (GFXTexHandle texObj : textureArray)
+   {
+      if (texObj.isValid())
+      {
+         // dynamic_cast<GFXGLTextureObject*>(texObj.getPointer())->
```
Comment (lukaspj) [inline Engine/source/gfx/gl/gfxGLTextureArray.cpp]: Remove comment

File: Engine/source/shaderGen/shaderOp.cpp [hunk 15]
```diff
@@ -80,6 +80,29 @@ void EchoOp::print( Stream &stream )
    WRITESTR( mStatement );
 }
 
+//**************************************************************************
+// Echo operation - deletes incoming statement!
```
Comment (lukaspj) [inline Engine/source/shaderGen/shaderOp.cpp]: Fix comment

**Review by lukaspj — COMMENTED** [review]
Need a look through the shadergen

**Comment by lukaspj** [comment]
Hmm come to think of it, I haven't tested with macros as they weren't working anyways, that probably need to be looked at before merging.

File: Engine/source/gfx/gl/gfxGLTextureArray.h [hunk 16]
```diff
@@ -0,0 +1,40 @@
+#ifndef _GFXD3D11TEXTUREARRAY_H_
+#define _GFXD3D11TEXTUREARRAY_H_
```
Comment (rextimmy) [inline Engine/source/gfx/gl/gfxGLTextureArray.h]: D3D11 header guards

File: Engine/source/gfx/gl/gfxGLTextureArray.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <dxgiformat.h>
```
Comment (rextimmy) [inline Engine/source/gfx/gl/gfxGLTextureArray.h]: more dx code

File: Engine/source/gfx/gl/gfxGLTextureArray.h [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+{
+public:
+   GFXGLTextureArray();
+
+   ~GFXGLTextureArray();
+
+   bool fromTextureArray(Vector<GFXTexHandle> textureArray) override;
+   void setToTexUnit(U32 tuNum) override;
+
+   void createResourceView(DXGI_FORMAT format, U32 numMipLevels, U32 usageFlags);
```
Comment (rextimmy) [inline Engine/source/gfx/gl/gfxGLTextureArray.h]: more dx code

File: Engine/source/gfx/gfxTextureArray.h [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+#endif
+
+
+class GFXTextureProfile;
+class GFXTextureObject;
+
+class GFXTextureArray : public StrongRefBase, public GFXResource
+{
+public:
+   virtual bool fromTextureArray(Vector<GFXTexHandle> textureArray) = 0;
```
Comment (rextimmy) [inline Engine/source/gfx/gfxTextureArray.h]: unnecessary copy happening here, pass the vector by reference (preferably const)

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+
+   D3D11_TEXTURE2D_DESC texArrayDesc;
+   texArrayDesc.Width = texElementDesc.Width;
+   texArrayDesc.Height = texElementDesc.Height;
+   texArrayDesc.MipLevels = texElementDesc.MipLevels;
+   texArrayDesc.ArraySize = mArraySize;
+   texArrayDesc.Format = GFXD3D11TextureFormat[format];
+   texArrayDesc.SampleDesc.Count = 1;
+   texArrayDesc.SampleDesc.Quality = 0;
+   texArrayDesc.Usage = D3D11_USAGE_DEFAULT;
```
Comment (rextimmy) [inline Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: D3D11_USAGE_DEFAULT can't be used with the cpu access flag D3D11_CPU_ACCESS_WRITE. See here https://docs.microsoft.com/en-us/windows/win32/api/d3d11/ne-d3d11-d3d11_cpu_access_flag

**Comment by lukaspj** [comment]
Thanks for the review @rextimmy !

File: Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp (hunk 20 above)
Comment (lukaspj) [reply Engine/source/gfx/D3D11/gfxD3D11TextureArray.cpp]: Turns out I could remove the CPU_ACCESS flag completely, nice catch!

**Comment by Azaezel** [comment]
should note this one's superseded by #424

**Comment by lukaspj** [comment]
In my tantrums I let #424 diverge from this so I will close this

**Pull request closed without merging**